### PR TITLE
iox-1732 implement bump allocator

### LINKED ISSUE #1732: Implement BumpAllocator
## Brief feature description

Currently we have (at least) two classes to allocate memory: `posix::Allocator` and `cxx::static_storage`, where `cxx::static_storage` is basically `posix::Allocator` but with internal storage. To have a more generic and reusable allocator, we can remove the internal storage from `cxx::static_storage`, merge it with `posix::Allocator` and rename it to `BumpAllocator` (as known in literature). This allocator class can use the `unintialized_array` (#1614) underneath.

For the `storable_function`, this means that the storage needs then to be moved to the `storable_function` and the `BumpAllocator` needs to be initialized with a pointer to this storage.

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR implements the `BumpAllocator`. Therefor the `Allocator` form the shared memory object directory is moved to `memory`, `deallocate` is implemented and the option to finalize the allocation is moved to the shared memory object. Additionally, the `static_storage` is moved to `iceoryx_dust` and a simple byte array is used in the `storable_function` instead.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1732

### CODE REVIEW

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl [hunk 1]
```diff
@@ -166,22 +172,28 @@ inline void swap(storable_function<Capacity, T>& f, storable_function<Capacity,
     f.swap(g);
 }
 
+template <typename T>
+void* allocate(byte_t* startAddress, uint64_t Capacity)
+{
+    // NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast, performance-no-int-to-ptr) required for low level pointer alignment
+    uint64_t alignment = alignof(T);
+    uint64_t alignedPosition = cxx::align(reinterpret_cast<uint64_t>(startAddress), alignment);
+    cxx::Expects(alignedPosition + sizeof(T) - reinterpret_cast<uint64_t>(startAddress) < Capacity
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl:181]: @elBoberido @FerdinandSpitzschnueffler

Now we have a regression. Previously it could be detected at compile time that the allocation will fail due to insufficient size for a callable (if used with some type that was too large). This was one reason for the static storage. Now we have checks at runtime for this, which leads to some issues which are just not necessary.

1.) Performance overhead (if the check is enabled, which it currently is)
2.) We may not know at compile time that we have insufficient space and cannot guarantee that it runs at compile time (with sufficient stack, but this is still required).

This is not good, we want to follow a fail early philosophy and you cannot really fail much earlier than compile time. Note that we could add the compile time check here also directly, as it was done in `static_storage`. However, I do not quite see the improvement. We shifted more of the storage responsibility to `function`, instead of abstracting it in something like `static_storage`.

**Review by MatthiasKillat — COMMENTED** [review]
Only reviewed the initial changes up to `cxx::function` changes. Will continue tomorrow.

File: iceoryx_posh/source/roudi/memory/mempool_collection_memory_block.cpp [hunk 2]
```diff
@@ -51,8 +51,10 @@ uint64_t MemPoolCollectionMemoryBlock::alignment() const noexcept
 
 void MemPoolCollectionMemoryBlock::onMemoryAvailable(cxx::not_null<void*> memory) noexcept
 {
-    posix::Allocator allocator(memory, size());
-    auto memoryManager = allocator.allocate(sizeof(mepoo::MemoryManager), alignof(mepoo::MemoryManager));
+    BumpAllocator allocator(memory, size());
+    auto allocationResult = allocator.allocate(sizeof(mepoo::MemoryManager), alignof(mepoo::MemoryManager));
+    cxx::Expects(!allocationResult.has_error());
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/roudi/memory/mempool_collection_memory_block.cpp:56]: Same pattern as before. If allocation fails we terminate. This may be tolerable here. It was not checked before and checking is better than continuing without resources.

It would be even better to continue, but this is probably not possible. The next best option is graceful termination, but we lack facilities to do so. This will be corrected soon. :-)

This pattern may occur more often and I will not further comment on it as there is no immediate solution here.

Just an observation, nothing to do for now (and no todo needed as searching for `Expects` is sufficient).

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    ::testing::Test::RecordProperty("TEST_ID", "4252ddcc-05d4-499f-ad7c-30bffb420e08");
+    iox::BumpAllocator sut(memory, memorySize);
+    auto allocationResult = sut.allocate(5, MEMORY_ALIGNMENT);
+    ASSERT_FALSE(allocationResult.has_error());
+    auto* bla = static_cast<uint8_t*>(allocationResult.value());
+
+    allocationResult = sut.allocate(5, MEMORY_ALIGNMENT);
+    ASSERT_FALSE(allocationResult.has_error());
+    auto* bla2 = static_cast<uint8_t*>(allocationResult.value());
+    EXPECT_THAT(bla2 - bla, Eq(8U));
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp:116]: That only shows that consecutive allocations are 8 bytes offset from each other but not that they are correctly aligned (more precisely it also assumes `bla2 > bla` which is OK for a bump allocator but not necessarily for others). Theoretically this condition does not need to hold, the gap could be larger and both addresses still correctly aligned. In this case some of the bytes are not used, but allocators do this all the time, depending on type.

The test should be changed to test the alignment of the addresses, meaning that they are multiples of whatever alignment is requested. The test name would be `allocationIsCorrectlyAligned`.

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+TEST_F(BumpAllocator_Test, allocateElementOfSizeZero)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp:119]:
```suggestion
TEST_F(BumpAllocator_Test, allocationOfZeroBytesTerminates)
```

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+{
+    ::testing::Test::RecordProperty("TEST_ID", "17caa50c-94bf-4a1d-a1ec-dfda563caa0b");
+    iox::BumpAllocator sut(memory, memorySize);
+
+    // @todo iox-#1613 remove EXPECT_DEATH
+    // NOLINTBEGIN(hicpp-avoid-goto, cppcoreguidelines-avoid-goto, cert-err33-c, cppcoreguidelines-pro-type-vararg,
+    // hiccpp-vararg)
+    EXPECT_DEATH(IOX_DISCARD_RESULT(sut.allocate(0, MEMORY_ALIGNMENT)), ".*");
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp:127]: We could also leave this as unspecified and hence not test it if we want to get rid of (death) tests. Allocating 0 bytes is not part of the contract.

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    // NOLINTEND(hicpp-avoid-goto, cppcoreguidelines-avoid-goto, cert-err33-c, cppcoreguidelines-pro-type-vararg,
+    // hiccpp-vararg)
+}
+
+TEST_F(BumpAllocator_Test, allocateAfterDeallocateWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "323fc1af-481f-4732-b7d3-fa32da389cef");
+    iox::BumpAllocator sut(memory, memorySize);
+    ASSERT_FALSE(sut.allocate(memorySize, 1).has_error());
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp:136]: Maybe extend to a more general case with multiple allocations. However, `deallocate` reclaims (resets) them all anyway by design.

We should test if the following allocation would be the start address (or more precisely the next aligned address after this) of the memory the bump allocator manages (i.e. `memory`).

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    sut.deallocate();
+
+    auto allocationResult = sut.allocate(memorySize, 1);
+    ASSERT_FALSE(allocationResult.has_error());
+    int* bla = static_cast<int*>(allocationResult.value());
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp:142]: We can fill the memory with something and check (but this was already done before, albeit without previous deallocation). It would be more important to check that we start with the "first" address before, i.e. the bump allocator is correctly reset.

Test name could be `deallocationResetsAllocator`.

**Review by MatthiasKillat — COMMENTED** [review]
The tests need some rework.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl:181]: I think we can improve this function so that it is compile time again.
1. `Capacity` becomes the second template argument
2. We add this `static_assert(sizeof(T) + alignof(T) - 1 <= Capacity, "type does not fit into storage");`

With this it should be possible to remove the `cxx::Expects`.

File: iceoryx_hoofs/memory/include/iox/bump_allocator.hpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    BumpAllocator& operator=(const BumpAllocator&) noexcept = delete;
+    BumpAllocator& operator=(BumpAllocator&&) noexcept = default;
+    ~BumpAllocator() noexcept = default;
+
+    /// @brief allocates on the memory supplied with the ctor
+    /// @param[in] size of the memory to allocate, must be greater than 0
+    /// @param[in] alignment of the memory to allocate
+    /// @return returns an expected containing a pointer to the allocated memory if memory is available, otherwise
+    /// BumpAllocatorError::OUT_OF_MEMORY
+    cxx::expected<void*, BumpAllocatorError> allocate(const uint64_t size, const uint64_t alignment) noexcept;
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/memory/include/iox/bump_allocator.hpp:52]: I am just putting it up for debate. And Rust has better builtin support for its result types, e.g. `?` and match which makes it less cumbersome to use

File: iceoryx_posh/source/mepoo/mem_pool.cpp [hunk 9]
```diff
@@ -41,18 +41,23 @@ constexpr uint64_t MemPool::CHUNK_MEMORY_ALIGNMENT;
...
     , m_minFree(numberOfChunks)
 {
     if (isMultipleOfAlignment(chunkSize))
     {
-        m_rawMemory = static_cast<uint8_t*>(chunkMemoryAllocator.allocate(
-            static_cast<uint64_t>(m_numberOfChunks) * m_chunkSize, CHUNK_MEMORY_ALIGNMENT));
-        auto memoryLoFFLi =
+        auto allocationResult = chunkMemoryAllocator.allocate(static_cast<uint64_t>(m_numberOfChunks) * m_chunkSize,
+                                                              CHUNK_MEMORY_ALIGNMENT);
+        cxx::Expects(!allocationResult.has_error());
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/mepoo/mem_pool.cpp:54]: @MatthiasKillat yes, this is handled during initialization and it should be ensured that enough memory is available.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl]: @MatthiasKillat yes, we replace the dependency on one class with another class but in total there is less code since one dependency had a large overlap with the one that replaced it.

File: iceoryx_hoofs/memory/source/bump_allocator.cpp [hunk 10]
```diff
@@ -15,32 +15,25 @@
...
 // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
-void* Allocator::allocate(const uint64_t size, const uint64_t alignment) noexcept
+cxx::expected<void*, BumpAllocatorError> BumpAllocator::allocate(const uint64_t size, const uint64_t alignment) noexcept
 {
-    cxx::Expects(size > 0);
-
-    cxx::Expects(
-        !m_allocationFinalized
-        && "allocate() call after finalizeAllocation()! You are not allowed to acquire shared memory chunks anymore");
+    cxx::ExpectsWithMsg(size > 0, "Cannot allocate memory of size 0");
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/memory/source/bump_allocator.cpp]: It is correct that `ExpectsWithMsg` causes termination (in the future only in debug mode). I removed it and return an error now.

File: iceoryx_binding_c/test/moduletests/test_notification_info.cpp [hunk 11]
```diff
@@ -102,7 +102,7 @@ class iox_notification_info_test : public Test
     static constexpr uint32_t CHUNK_SIZE = 128U;
     static constexpr size_t MEMORY_SIZE = 1024 * 1024 * 100;
     uint8_t m_memory[MEMORY_SIZE];
-    Allocator m_memoryAllocator{m_memory, MEMORY_SIZE};
+    BumpAllocator m_memoryAllocator{m_memory, MEMORY_SIZE};
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_binding_c/test/moduletests/test_notification_info.cpp:105]: I created #1849

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp (hunk 5 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp]: Since `BumpAllocator::allocate` returns now an `expected`, allocating 0 bytes is tested without an death test in `AllocateFailsWithZeroSize`.

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    *bla = 123;
+    EXPECT_THAT(*bla, Eq(123));
+}
+
+TEST_F(BumpAllocator_Test, allocateEverythingWithMultipleElements)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "21d0fa61-54f9-41a0-8e53-e3448784497b");
+    iox::BumpAllocator sut(memory, memorySize);
+    for (size_t i = 0; i < memorySize; i += 32)
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp]: Please see `AllocateCompleteMemoryWithEquallySizedChunksWorks` and `AllocateCompleteMemoryWithDifferentSizedChunksWorks`.

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    int* bla = static_cast<int*>(allocationResult.value());
+    *bla = 123;
+    EXPECT_THAT(*bla, Eq(123));
+}
+
+TEST_F(BumpAllocator_Test, allocateEverythingWithSingleElement)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f2e1085b-08fe-4b08-b022-0385b5a53fca");
+    iox::BumpAllocator sut(memory, memorySize);
+    auto allocationResult = sut.allocate(memorySize, 1);
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp]: I reworked all the `BumpAllocator` tests. For this test, please see `AllocateCompleteMemoryAndStoreDataWorks`.

File: iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+    void* memory{nullptr};
+    size_t memorySize = 10016;
+};
+
+TEST_F(BumpAllocator_Test, allocateOneSmallElement)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f689e95c-5743-4370-93f0-8a23b909c75a");
+    iox::BumpAllocator sut(memory, memorySize);
+    auto allocationResult = sut.allocate(sizeof(int), MEMORY_ALIGNMENT);
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_memory_bump_allocator.cpp]: Please see `AllocationIsCorrectlyAligned` and `AllocateSmallMemoryChunkAndStoreDataWorks`.

File: iceoryx_posh/source/roudi/memory/mempool_collection_memory_block.cpp (hunk 2 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/source/roudi/memory/mempool_collection_memory_block.cpp:56]: Just for info: the `Expects` was hidden inside `Allocator::allocate` before.

File: iceoryx_hoofs/test/moduletests/test_posix_shared_memory_object.cpp [hunk 15]
```diff
@@ -54,6 +54,8 @@ class SharedMemoryObject_Test : public Test
         // hiccpp-vararg)
         internal::CaptureStderr();
     }
+
+    static constexpr uint64_t MEMORY_ALIGNMENT{8};
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_posix_shared_memory_object.cpp:38]: Confirmed.

File: iceoryx_hoofs/memory/include/iox/bump_allocator.hpp [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+    /// @brief mark the memory as unused
+    void deallocate() noexcept;
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/memory/include/iox/bump_allocator.hpp:57]: I tend to keep the current name.

File: iceoryx_hoofs/memory/include/iox/bump_allocator.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/memory/include/iox/bump_allocator.hpp:54]: I tend to keep the return type, also added another allocation error.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl [hunk 17]
```diff
@@ -166,22 +172,28 @@ inline void swap(storable_function<Capacity, T>& f, storable_function<Capacity,
     f.swap(g);
 }
 
+template <typename T>
+void* allocate(byte_t* startAddress, uint64_t Capacity)
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl]: Changed the name to `safeAlign`.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl]: @MatthiasKillat @elfenpiff I added the compile time check.

**Review by elfenpiff — DISMISSED** [review]

File: iceoryx_dust/include/iceoryx_dust/internal/cxx/static_storage.hpp [hunk 18]
```diff
@@ -14,8 +14,8 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#ifndef IOX_HOOFS_STATIC_STORAGE_HPP
-#define IOX_HOOFS_STATIC_STORAGE_HPP
+#ifndef IOX_DUST_STATIC_STORAGE_HPP
+#define IOX_DUST_STATIC_STORAGE_HPP
```
Comment (MatthiasKillat) [inline iceoryx_dust/include/iceoryx_dust/internal/cxx/static_storage.hpp:19]: @FerdinandSpitzschnueffler We may consider fixing the last 2 remaining Axivion warnings right here, I can add a commit myself. It would be the C array usage in this file (questionable rule but OK).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.hpp [hunk 19]
```diff
@@ -155,7 +154,9 @@ class storable_function<Capacity, signature<ReturnType, Args...>> final
   private:
     operations m_operations; // operations depending on type-erased callable (copy, move, destroy)
 
-    StorageType m_storage;                              // storage for the callable
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : safe access is guaranteed since the c-array is wrapped inside the storable_function
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.hpp:157]: @elBoberido @FerdinandSpitzschnueffler I still think we need the storage abstraction in the future to avoid comments like this etc. I would like to distinguish between allocator and storage as different concepts.

Storage either magically exists or is provided by some other mechanism. In the end it abstracts memory, often but not necessarily contiguous. In a way it wraps memory but abstracts the access and other functionality.

An allocator is a means to get memory, or maybe storage. It can serve multiple allocations and internally generally does not need to manage contiguous memory.

The reason for storage abstraction is also to move to a more general concept where we decouple the storage and the object logic. Anyway, we can keep it for now but I do not see the benefit. Now the function implementation is more tightly coupled again to the storage concept and implements the core functionality (now called `safeAlign`).

Considerations for future refactoring. Resolve.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl [hunk 20]
```diff
@@ -309,14 +311,14 @@ template <uint64_t Capacity, typename ReturnType, typename... Args>
 template <typename T>
 inline constexpr uint64_t storable_function<Capacity, signature<ReturnType, Args...>>::required_storage_size() noexcept
 {
-    return StorageType::template allocation_size<T>();
+    return sizeof(T) + alignof(T) - 1;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl:314]: This could be made more exact if the alignment of storage is taken in account (may safe a little memory). But this worst case estimate is correct, just not tight in general. It assumes that we need to maximally realign by `alignof(T) - 1`, but generally it will be less.

It would look like the `align` method we refactored in helplets to get an aligned address from the start of storage and the look whether there is space for `sizeof(T)` from this address onward.

This is why it was previously part of the storage and hence knew about the alignment directly (and another reason for the abstraction). I consider reworking the storage concepts in the future to better reflect this, as it existed before it was functional but not fully developed.

TLDR: can be optimized but does not have to be right now. A change would need to take the starting address of `m_storage` into account.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/shared_memory_object.hpp [hunk 21]
```diff
@@ -40,6 +40,14 @@ enum class SharedMemoryObjectError
     INTERNAL_LOGIC_FAILURE,
 };
 
+enum class SharedMemoryAllocationError
+{
+    REQUESTED_MEMORY_AFTER_FINALIZED_ALLOCATION,
+    NOT_ENOUGH_MEMORY,
+    REQUESTED_ZERO_SIZED_MEMORY
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/shared_memory_object.hpp:47]: Requesting zero memory should be a misuse case. We do not have to change it now as it is a more fundamental issue, but I think we should not mix errors that can happen during correct use and incorrect use. Note that some allocator implementations just treat requesting zero bytes as requesting one byte by convention, but we should not do this IMO.

I have some notion of this in the new error handling and I think the distinction is needed to develop better/safe code.

Nothing to do now.

File: iceoryx_hoofs/memory/include/iox/bump_allocator.hpp [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+#include <cstdint>
+
+namespace iox
+{
+
+enum class BumpAllocatorError
+{
+    OUT_OF_MEMORY,
+    REQUESTED_ZERO_SIZED_MEMORY
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/memory/include/iox/bump_allocator.hpp:31]: See my comment about misuse cases vs. legitimate errors. No change required now but fundamentally there should be a distinction.

File: iceoryx_hoofs/source/posix_wrapper/shared_memory_object.cpp [hunk 23]
```diff
@@ -166,17 +166,36 @@ SharedMemoryObject::SharedMemoryObject(SharedMemory&& sharedMemory,
...
 
-void* SharedMemoryObject::allocate(const uint64_t size, const uint64_t alignment) noexcept
+cxx::expected<void*, SharedMemoryAllocationError> SharedMemoryObject::allocate(const uint64_t size,
+                                                                               const uint64_t alignment) noexcept
 {
-    return m_allocator.allocate(size, alignment);
+    if (size == 0)
+    {
+        IOX_LOG(WARN) << "Cannot allocate memory of size 0.";
+        return cxx::error<SharedMemoryAllocationError>(SharedMemoryAllocationError::REQUESTED_ZERO_SIZED_MEMORY);
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/source/posix_wrapper/shared_memory_object.cpp:175]: IMO this should be an assert like check, it is a violation of the precondition size>0.

Precondition violations indicate bugs and should never lead to error propagation but (as graceful as possible) termination and a log entry indicating the bug.

We can solve this later while reworking how we deal with errors though.

File: iceoryx_hoofs/source/posix_wrapper/shared_memory_object.cpp [hunk 24]
```diff
@@ -166,17 +166,36 @@ SharedMemoryObject::SharedMemoryObject(SharedMemory&& sharedMemory,
...
+    }
+    if (m_allocationFinalized)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/source/posix_wrapper/shared_memory_object.cpp:177]: This error condition in turn is a bit on the fence. Arguably one can request the system designer to ensure to never call it after finalization (if he controls finalization as well). On the other hand it is much harder to do when finalization is controlled elsewhere. So this is a legitimate error case that needs to be communicated somehow.

Same with being out of memory at runtime. Controllable to some degree, but not in all usage scenarios.

Resolve me.

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp [hunk 25]
```diff
@@ -252,9 +253,16 @@ TEST_F(function_test, ConstructionFromAnotherFunctionIsCallable)
 TEST_F(function_test, FunctionStateIsIndependentOfSource)
 {
     ::testing::Test::RecordProperty("TEST_ID", "8302046f-cd6a-4527-aca6-3e6408f87a6b");
-    constexpr uint32_t INITIAL_STATE = 73U;
-    static_storage<1024U> storage;
-    auto* p = storage.allocate<Functor>();
+    constexpr uint32_t INITIAL_STATE = 73;
+    constexpr uint32_t MEMORY_SIZE = 1024;
+
+    // NOLINTNEXTLINE(hicpp-no-malloc, cppcoreguidelines-no-malloc) low-level memory management for testing purpose
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:259]: @elBoberido 

I was not aware that we enforce static analysis rules in test code as well. I think we should not (we already established that we use certain constructs in tests that are not allowed in library code, such as dynamic memory). If we do enforce the rules we need the comment of course.

**Review by MatthiasKillat — APPROVED** [review]
The Bumballocator is fine IMO (apart from defining zero allocation as an error).

I still do not see the benefit of moving the storage and reimplementing the core functionality in the `function` implementation itself now. Also, since some changes before to modularity/policies of `storable_function`, there is no need anymore to have different names `storable_function` and `function` (only function suffices now). The idea was that a `storable_function` with `static_storage` policy is the general purpose function wrapper we use - `function`. But there can be other ones, with different storage policy. This is obsolete now due to the many changes but the idea is still sound I wuld argue.

Since this is not a major part of `BumpAllocator` refactoring we should not change this here and now, but I recommend we reconsider the storage concept in the future.

**Comment by MatthiasKillat** [comment]
@FerdinandSpitzschnueffler There are conflicts unfortunately. Will approve again once they are resolved.

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp (hunk 25 above)
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:259]: @MatthiasKillat we enforce a subset of what we do for production code. In the past we also had some issues in the tests which at some point resulted in weird test failures. If the current list of checks is too strict for the tests, we can of course talk about relaxing them.

**Review by MatthiasKillat — APPROVED** [review]

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp (hunk 25 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:259]: I see. Dynamic memory allocation might be one of them, since we allow `new` and `delete` for objects created in tests `malloc` and `free` should be allowed for similar reasons without a comment.

**Review by elfenpiff — APPROVED** [review]
